keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,220 @@
+    Exec clientcmd.ExecConfig
+}
+```
+
+The existing `ExecCredentialSpec` and `ExecCredentialStatus` will add support for registry credentials.  In the in-tree implementation, the docker keyring,
+which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.
+
+```go
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+    // +optional
+    Response *Response `json:"response,omitempty"`
+
+    // +optional
+    Interactive bool `json:"interactive,omitempty"`
+
+    // Adding field:
+    // Image is used when passed to registry credential providers to as part of an image pull
+    // +optional
+    Image string `json:"image,omitempty"`
+}
+
+// ExecCredentialStatus holds credentials for kube runtime to use for image pulls.
+type ExecCredentialStatus struct {
+    // +optional
+    ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+    Token string `json:"token,omitempty"`
+    ClientCertificateData string `json:"clientCertificateData,omitempty"`
+    ClientKeyData string `json:"clientKeyData,omitempty"`
+
+    // Adding field:
+    // AuthConfig information for connecting to a Registry
+    // +optional
+    AuthConfig *RegistryAuthConfig `json:"authConfig,omitempty"`
+}
+
+// Adding type:
+// RegistryAuthConfig contains authorization information for connecting to a Registry
+// This type mirrors "github.com/docker/docker/api/types.AuthConfig"
+type RegistryAuthConfig struct {
+    Username string `json:"username,omitempty"`
+    Password string `json:"password,omitempty"`
+    Auth     string `json:"auth,omitempty"`
+
+    // Email is an optional value associated with the username.
+    // This field is deprecated and will be removed in a later
+    // version of docker.
+    Email string `json:"email,omitempty"`
+
+    ServerAddress string `json:"serveraddress,omitempty"`
+
+    // IdentityToken is used to authenticate the user and get
+    // an access token for the registry.
+    IdentityToken string `json:"identitytoken,omitempty"`
+
+    // RegistryToken is a bearer token to be sent to a registry
+    RegistryToken string `json:"registrytoken,omitempty"`
+}
+
+```
+
+### Alternatives Considered
+
+#### API Server Proxy
+
+The API server would act as a proxy to an external container registry credential provider that may support multiple cloud providers. The credential provider service will return container runtime compatible responses of the type currently used by the credential provider infrastructure within the kubelet along with credential expiration information to allow the API server to cache credential responses for a period of time.
+
+This limits the cloud-specific privileges required for each node for the purpose of fetching credentials. Centralized caching helps to avoid cloud-specific rate limits for credential acquisition by consolidating that credential acquisition within the API server.
```

Comment:
> Centralized caching helps to avoid cloud-specific rate limits for credential acquisition by consolidating that credential acquisition within the API server.

Also, this is unrelated to the stated goals of this KEP.